File: src/dialogue_bridge/router/skills.py
"""Skills router — global catalog + per-user pool + per-(user, agent) selection.

Three tiers:

- ``GET /v1/skills`` returns the global admin-curated catalog.
- ``GET / POST / DELETE /v1/users/{user_id}/skills[/...]`` manage the
  user's personal pool — add globals by name, create owned customs,
  remove (cascades to per-agent assignments).
- ``GET / PUT / DELETE /v1/users/{user_id}/agents/{agent_id}/skills[/{name}]``
  manage which pool skills are assigned to a deep agent. The agents service
  resolves the source folder via the user's manifest at assignment time.

The bridge proxies to the agents service (which owns every volume) and
caches reads in Redis with short TTLs + explicit invalidation on every
mutation.

The selection state is **not stored in Postgres** — the on-disk filesystem
under ``$SKILLS_REGISTRY_USERS_ROOT/<user_id>/`` and
``$AGENTS_FILESYSTEM_ROOT/<user_id>/<agent_slug>/skills/`` IS the source of
truth.
"""
from __future__ import annotations

from typing import List

from fastapi import APIRouter, Depends, Query, status
from core.logging import get_logger, set_context

from core.auth.session import AuthUser, require_csrf_protection, require_current_user
from core.security.rate_limit import skill_upload_rate_limit
from schemas import (
    CustomSkillCreateRequest,
    Skill,
    UserSkill,
    UserSkillDetail,
)
from utils import validate_userId
from utils.skills import (
    add_global_skill_to_user_pool,
    create_custom_skill_in_pool,
    disable_user_agent_skill,
    enable_user_agent_skill,
    get_user_agent_skills,
    get_user_skill_detail,
    list_skills,
    list_user_skills,
    remove_skill_from_user_pool,
)

router = APIRouter()
logger = get_logger(__name__)


@router.get("", response_model=List[Skill], status_code=status.HTTP_200_OK)
async def get_skills(
    bypass_redis: bool = Query(
        default=False,
        description=(
            "When true, skip the Redis cache, fetch fresh from the agents "
            "service (forcing its in-memory manifest rebuild), and upsert "
            "the bridge cache with the response. Used by the manual "
            "'refresh' button. A normal page refresh leaves this false."
        ),
    ),
    _: AuthUser = Depends(require_current_user),
) -> List[Skill]:
    """Return every skill in the global catalog (admin-curated). Requires a valid session."""
    payload = await list_skills(bypass_cache=bypass_redis)
    logger.info(
        "skills_listed",
        "Served global skills catalog to UI",
        count=len(payload),
        bypass_redis=bypass_redis,
    )
    return [Skill.model_validate(item) for item in payload]


# ---------------------------------------------------------------------------
# Per-user pool
# ---------------------------------------------------------------------------
@router.get(
    "/users/{user_id}",
    response_model=List[UserSkill],
    status_code=status.HTTP_200_OK,
)
async def get_user_pool(
    user_id: str,
    bypass_redis: bool = Query(default=False),
    _: AuthUser = Depends(validate_userId),
) -> List[UserSkill]:
    """Return the user's skill pool manifest (no content)."""
    set_context(user_id=user_id)
    payload = await list_user_skills(user_id=user_id, bypass_cache=bypass_redis)
    logger.info(
        "user_pool_listed",
        "Served user skill pool",
        user_id=user_id,
        count=len(payload),
        bypass_redis=bypass_redis,
    )
    return [UserSkill.model_validate(item) for item in payload]


@router.get(
    "/users/{user_id}/{skill_name}",
    response_model=UserSkillDetail,
    status_code=status.HTTP_200_OK,
)
async def get_user_skill(
    user_id: str,
    skill_name: str,
    _: AuthUser = Depends(validate_userId),
) -> UserSkillDetail:
    """Return a single user-pool skill with its SKILL.md content (no cache)."""
    set_context(user_id=user_id)
    payload = await get_user_skill_detail(user_id=user_id, skill_name=skill_name)
    return UserSkillDetail.model_validate(payload)


@router.post(
    "/users/{user_id}/global/{skill_name}",
    status_code=status.HTTP_204_NO_CONTENT,
)
async def add_global_to_pool(
    user_id: str,
    skill_name: str,
    _: AuthUser = Depends(validate_userId),
    __: None = Depends(require_csrf_protection),
) -> None:
    """Append a global-skill reference to the user's pool."""
    set_context(user_id=user_id)
    await add_global_skill_to_user_pool(user_id=user_id, skill_name=skill_name)


@router.post(
    "/users/{user_id}/custom",
    response_model=UserSkill,
    status_code=status.HTTP_201_CREATED,
    # Writes multi-file folders onto the agents-service disk — per-user ceiling.
    dependencies=[Depends(skill_upload_rate_limit)],
)
async def create_custom_skill(
    user_id: str,
    payload: CustomSkillCreateRequest,
    _: AuthUser = Depends(validate_userId),
    __: None = Depends(require_csrf_protection),
) -> UserSkill:
    """Create a user-owned custom skill in the pool."""
    set_context(user_id=user_id)
    body = await create_custom_skill_in_pool(
        user_id=user_id, payload=payload.model_dump()
    )
    return UserSkill.model_validate(body)


@router.delete(
    "/users/{user_id}/{skill_name}",
    status_code=status.HTTP_204_NO_CONTENT,
)
async def remove_from_pool(
    user_id: str,
    skill_name: str,
    _: AuthUser = Depends(validate_userId),
    __: None = Depends(require_csrf_protection),
) -> None:
    """Remove a skill from the user's pool, cascading to per-agent assignments."""
    set_context(user_id=user_id)
    await remove_skill_from_user_pool(user_id=user_id, skill_name=skill_name)


@router.get(
    "/users/{user_id}/agents/{agent_id}",
    response_model=List[str],
    status_code=status.HTTP_200_OK,
)
async def get_enabled_skills_for_user_agent(
    user_id: str,
    agent_id: str,
    _: AuthUser = Depends(validate_userId),
) -> List[str]:
    """Return the names of skills enabled for this (user, agent) pair.

    Read-through Redis cache with a short TTL; mutations always invalidate
    the matching key, so the TTL is just a safety net.
    """
    set_context(user_id=user_id, agent_id=agent_id)
    skills = await get_user_agent_skills(user_id=user_id, agent_id=agent_id)
    logger.info(
        "user_agent_skills_served",
        "Served per-(user, agent) skill selection",
        count=len(skills),
    )
    return skills


@router.put(
    "/users/{user_id}/agents/{agent_id}/{skill_name}",
    status_code=status.HTTP_204_NO_CONTENT,
)
async def enable_skill_for_user_agent(
    user_id: str,
    agent_id: str,
    skill_name: str,
    _: AuthUser = Depends(validate_userId),
    __: None = Depends(require_csrf_protection),
) -> None:
    """Enable ``skill_name`` for this (user, agent) pair.

    Proxies to the agents service which copies the registry's skill
    directory into the per-user filesystem. Idempotent: re-enabling an
    already-enabled skill is a no-op upstream.
    """
    set_context(user_id=user_id, agent_id=agent_id)
    await enable_user_agent_skill(
        user_id=user_id, agent_id=agent_id, skill_name=skill_name
    )


@router.delete(
    "/users/{user_id}/agents/{agent_id}/{skill_name}",
    status_code=status.HTTP_204_NO_CONTENT,
)
async def disable_skill_for_user_agent(
    user_id: str,
    agent_id: str,
    skill_name: str,
    _: AuthUser = Depends(validate_userId),
    __: None = Depends(require_csrf_protection),
) -> None:
    """Disable ``skill_name`` for this (user, agent) pair.

    Proxies to the agents service which removes the skill's directory from
    the per-user filesystem. Idempotent: disabling a not-enabled skill is
    a no-op upstream.
    """
    set_context(user_id=user_id, agent_id=agent_id)
    await disable_user_agent_skill(
        user_id=user_id, agent_id=agent_id, skill_name=skill_name
    )

File: magenticx/agentic_ui/src/shared/lib/api/skills.ts
/**
 * Skills API — the global registry, the per-(user, agent) enablement toggles,
 * and the user's personal skill pool (globals adopted + custom uploads).
 */
import type { CustomSkillCreatePayload, Skill, UserSkill, UserSkillDetail } from "../types";
import { requestJson, requestVoid } from "../http";
import { PROXY_LIMIT_MB } from "../uploadGuards";
import {
  SkillListSchema,
  StringListSchema,
  UserSkillDetailSchema,
  UserSkillListSchema,
  UserSkillSchema,
} from "../schemas";
import { SKILLS_BASE_PATH } from "./paths";

// Fetch the central skills registry. Used by the bootstrap path
// (auth handlers + auth rehydrate hook) the same way as ``getAgents`` and
// ``getTools`` — fetched on every page refresh, with an IndexedDB snapshot
// for instant paint while the request is in flight. Pass ``{ bypassRedis:
// true }`` to force the bridge to skip its Redis cache and re-fetch from the
// agents service — that path also upserts the cache so the next normal
// request benefits from the new data. Only the manual refresh button uses
// the bypass; everything else (login, page refresh) leaves it false so the
// cache absorbs the load.
export async function getSkills(options?: { bypassRedis?: boolean }): Promise<Skill[]> {
  const url = options?.bypassRedis ? `${SKILLS_BASE_PATH}?bypass_redis=true` : SKILLS_BASE_PATH;
  return requestJson(url, {
    schema: SkillListSchema,
    fallbackMessage: "Failed to fetch skills",
  });
}

// Fetch enabled skills for a (user, agent) pair.
// Returns a plain string[] of skill names. The bridge reads through a Redis
// cache; mutations invalidate the cache and the next GET re-fetches from the
// agents service (which is authoritative — the on-disk directory IS the state).
export async function getUserAgentSkills(userId: string, agentId: string): Promise<string[]> {
  const url = `${SKILLS_BASE_PATH}/users/${encodeURIComponent(userId)}/agents/${encodeURIComponent(agentId)}`;
  return requestJson(url, {
    schema: StringListSchema,
    fallbackMessage: "Failed to fetch user-agent skills",
  });
}

// Enable a skill for the (user, agent) pair. 204 on success.
export async function enableUserAgentSkill(
  userId: string,
  agentId: string,
  skillName: string,
): Promise<void> {
  const url = `${SKILLS_BASE_PATH}/users/${encodeURIComponent(userId)}/agents/${encodeURIComponent(agentId)}/${encodeURIComponent(skillName)}`;
  await requestVoid(url, {
    method: "PUT",
    csrf: true,
    fallbackMessage: `Failed to enable skill ${skillName}`,
  });
}

// Disable a skill for the (user, agent) pair. 204 on success (idempotent).
export async function disableUserAgentSkill(
  userId: string,
  agentId: string,
  skillName: string,
): Promise<void> {
  const url = `${SKILLS_BASE_PATH}/users/${encodeURIComponent(userId)}/agents/${encodeURIComponent(agentId)}/${encodeURIComponent(skillName)}`;
  await requestVoid(url, {
    method: "DELETE",
    csrf: true,
    fallbackMessage: `Failed to disable skill ${skillName}`,
  });
}

// ---------------------------------------------------------------------------
// Per-user skill pool (the user's personal registry of globals + customs)
// ---------------------------------------------------------------------------

// Fetch the user's pool manifest entries (no SKILL.md content).
//
// No bypass option: the bridge owns the pool in chat_db and re-imports from the
// agents service by itself whenever what it holds is missing or incomplete, so
// there is nothing a caller could usefully force.
export async function getMySkills(userId: string): Promise<UserSkill[]> {
  const url = `${SKILLS_BASE_PATH}/users/${encodeURIComponent(userId)}`;
  return requestJson(url, {
    schema: UserSkillListSchema,
    fallbackMessage: "Failed to fetch my skills",
  });
}

// Fetch a single user-pool skill with its SKILL.md body. Used when the user
// expands a card in the My skills view.
export async function getMySkillDetail(
  userId: string,
  skillName: string,
): Promise<UserSkillDetail> {
  const url = `${SKILLS_BASE_PATH}/users/${encodeURIComponent(userId)}/${encodeURIComponent(skillName)}`;
  return requestJson(url, {
    schema: UserSkillDetailSchema,
    fallbackMessage: `Failed to fetch skill detail ${skillName}`,
  });
}

// Append a global-catalog skill into the user's pool. 204 on success. 404 if
// the skill isn't in the global catalog; 409 if it's already in the pool.
export async function addGlobalSkillToPool(userId: string, skillName: string): Promise<void> {
  const url = `${SKILLS_BASE_PATH}/users/${encodeURIComponent(userId)}/global/${encodeURIComponent(skillName)}`;
  await requestVoid(url, {
    method: "POST",
    csrf: true,
    fallbackMessage: `Failed to add global skill ${skillName} to pool`,
  });
}

// Create a user-owned custom skill in the pool. 201 with the new manifest
// entry on success; 409 on name collision with global or own pool.
export async function createCustomSkill(
  userId: string,
  payload: CustomSkillCreatePayload,
): Promise<UserSkill> {
  const url = `${SKILLS_BASE_PATH}/users/${encodeURIComponent(userId)}/custom`;
  return requestJson(url, {
    method: "POST",
    csrf: true,
    body: payload,
    schema: UserSkillSchema,
    errorMessages: {
      413: `This skill is too large for the server (limit ${PROXY_LIMIT_MB} MB including base64 overhead). Use smaller files.`,
    },
    fallbackMessage: "Failed to create custom skill",
  });
}

// Remove a skill from the user's pool. Cascades on the agents service —
// also removes the skill from every per-(user, agent) assignment folder.
export async function removeSkillFromPool(userId: string, skillName: string): Promise<void> {
  const url = `${SKILLS_BASE_PATH}/users/${encodeURIComponent(userId)}/${encodeURIComponent(skillName)}`;
  await requestVoid(url, {
    method: "DELETE",
    csrf: true,
    fallbackMessage: `Failed to remove skill ${skillName} from pool`,
  });
}
